### PR TITLE
Swift plugin

### PR DESCRIPTION
**Pull request by sirsean**
This adds an Onyx plugin for writing messages to a Swift cluster. It currently only connects to a local mimic instance, but that should be merely a configuration change to switch it over to real Cloud Files.

Because this is closer to our real data store, the temporary MySQL database has been removed. We no longer start it in docker-compose and we no longer attempt to connect to it. All SQL-related code has been removed.

### CODE REVIEW

File: test/desdemona/plugins/swift_test.clj [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    (is (= cloud-files got))))
+
+(deftest get-token-test
+  (let [token {:id 1}
+        auth-response {:access {:token token}}
+        got (swift/get-token auth-response)]
+    (is (= token got))))
+
+(deftest authenticate-test
+  "This test verifies that the authenticate function makes a request to the
```
Comment (derwolfe) [inline test/desdemona/plugins/swift_test.clj]: @sirsean `lein eastwood` should stop complaining if you put your test descriptions inside of the `testing` macro.
```clojure
(deftest authenticate-test
  (testing "This test verifies..."
    (the-test-body)))
```
Here is the source for the testing macro: https://github.com/clojure/clojure/blob/d5708425995e8c83157ad49007ec2f8f43d8eac8/src/clj/clojure/test.clj#L596-L602

File: README.md [hunk 2]
```diff
@@ -66,19 +66,6 @@ Make sure you create the Kafka topic by connecting to the producer:
...
-script/connect_mysql.sh
-```
-
-Run this SQL, which is available in `resources/table.sql`:
-
-```
-use logs;
-CREATE TABLE logLines (id int primary key auto_increment, line text);
-```
-
```
Comment (lvh) [inline README.md]: :100:

File: docker-compose.yml [hunk 3]
```diff
@@ -3,11 +3,11 @@ peer:
   links:
     - zookeeper:zk
     - kafka:kafka
-    - db:db
     - syslog-ng:syslog-ng
+    - mimic:mimic
   environment:
     ONYX_ID: 1
-    N_PEERS: 6
+    N_PEERS: 8
```
Comment (lvh) [inline docker-compose.yml]: This was a problem, right? We should have a comment in here about the correct number of peers :)

File: script/mimic [hunk 4]
```diff
@@ -0,0 +1 @@
+Subproject commit 203f85db5de57703e7100a00636060f3095dda29
```
Comment (lvh) [inline script/mimic]: uh, huh? is mimic a submodule?

File: src/desdemona/jobs/sample_submit_job.clj [hunk 5]
```diff
@@ -27,15 +27,13 @@
...
+                                        :onyx/batch-timeout batch-timeout
+                                        :swift/auth-url "http://mimic:8900/identity/v2.0/tokens"
+                                        :swift/username "mimic"
+                                        :swift/api-key "12345"})
         (add-logging :read-lines)
-        (add-logging :write-lines))))
+        (add-logging :write-swift))))
 
 (defn -main [& args]
   (let [config (read-config (clojure.java.io/resource "config.edn") {:profile :dev})
```
Comment (lvh) [inline src/desdemona/jobs/sample_submit_job.clj]: Please require clojure.java.io and use io/resource instead

File: src/desdemona/plugins/swift.clj [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+  [auth-response]
+  (-> auth-response :access :token))
+
+(defn authenticate
+  "Authenticate with Cloud Files and return the authentication response. This
+  response will include an auth token that can be used for subsequent requests,
+  as well as a service catalog that will indicate the URL to use for, for example,
+  Cloud Files."
+  [auth-url username api-key]
+  (let [body (json/generate-string {"auth" {"RAX-KSKEY:apiKeyCredentials" {"username" username
```
Comment (lvh) [inline src/desdemona/plugins/swift.clj]: So, this module is called swift, but it looks like it has a bunch of Rackspace-specific stuff.

File: src/desdemona/plugins/swift.clj [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+  (write-batch
+    [_ {:keys [onyx.core/results]}]
+    (let [segments (map :message (mapcat :leaves (:tree results)))]
+      (when (not-empty segments)
+        (let [auth-response (authenticate auth-url username api-key)
+              auth-token (-> auth-response get-token :id)
+              cloud-files-url (-> auth-response get-cloud-files :endpoints first :public-url)
+              container-name (calculate-container-name)
+              file-name (calculate-file-name)]
+          (println "Writing" (count segments) "segments to" container-name file-name)
```
Comment (lvh) [inline src/desdemona/plugins/swift.clj]: Please use taoensso.timbre/info for logging instead

**Comment by reaperhulk** [comment]
jenkins, retest this please

File: test/desdemona/plugins/swift_test.clj [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+(deftest authenticate-test
+  (testing "This test verifies that the authenticate function makes a request to the
+           URL specified by the auth-url parameter, and includes the username and api-key
+           in the expected JSON format."
+    (http/with-fake-routes
+      {"http://authurl.com" {:post (fn [req]
+                                     (let [body (json/parse-stream (io/reader (req :body)) ->kebab-case-keyword)
+                                           username (-> body :auth :rax-kskey:api-key-credentials :username)
+                                           api-key (-> body :auth :rax-kskey:api-key-credentials :api-key)]
```
Comment (lvh) [inline test/desdemona/plugins/swift_test.clj]: Nitpick: this is an example of a place you might want to consider a `juxt`; a la:

`[username api-key] (-> ... ((juxt :username :api-key)))`

File: script/mimic (hunk 4 above)
Comment (sirsean) [inline script/mimic]: In order to run it inside Docker, I checked out the mimic repo and have it living here. Does git magically make that a submodule?

File: src/desdemona/plugins/swift.clj (hunk 6 above)
Comment (sirsean) [inline src/desdemona/plugins/swift.clj]: True. Do you think we should abstract it such that we pass in Rackspace-specific URLs & auth info, or should we name it something that describes that it's about Rackspace?

Comment (lvh) [inline src/desdemona/plugins/swift.clj]: Eh, I dunno. We don't have to do it right now, but we should probably track that, especially if we won't maintain this inside Desdemona.

File: test/desdemona/plugins/swift_test.clj (hunk 8 above)
Comment (sirsean) [inline test/desdemona/plugins/swift_test.clj]: Not sure why GitHub isn't picking up this change, but I switched this to `juxt` as noted. Thanks.

Comment (reaperhulk) [inline test/desdemona/plugins/swift_test.clj]: It did pick it up, you're commenting on an outdated diff?

File: project.clj [hunk 9]
```diff
@@ -18,6 +18,10 @@
                  [byte-streams "0.2.1"]
                  [camel-snake-kebab "0.3.2"]
 
+                 [clj-http "2.1.0"]
+                 [clj-http-fake "1.0.2"]
+                 [clj-time "0.11.0"]
+
```
Comment (lvh) [inline project.clj]: Was there any particular reason for using clj-http over e.g. aleph?

Comment (sirsean) [inline project.clj]: Didn't know about aleph. (I could look at my Google history but I think this was at the top of "Clojure HTTP request library" or something.)

Should I switch it over? (And does aleph have a request mocking library that goes along with it, because I think clj-http-fake is just for clj-http.)

File: resources/test/index.html [hunk 10]
```diff
@@ -0,0 +1,2 @@
+<!DOCTYPE html>
+<html><html><head><title>Desdemona</title><meta charset="utf-8"><meta content="width=device-width, initial-scale=1" name="viewport"><link href="/css/main.css" rel="stylesheet" type="text/css"></head><body><div id="app"></div><script src="/js/main.js" type="text/javascript"></script></body></html></html>
```
Comment (lvh) [inline resources/test/index.html]: Why is this here?

Comment (lvh) [inline resources/test/index.html]: (I think the answer to that question may be to merge forward)

File: script/syslog-ng/Dockerfile [hunk 11]
```diff
@@ -4,16 +4,22 @@ MAINTAINER Andras Mitzki <andras.mitzki@balabit.com>
 RUN apt-get update -qq && apt-get install -y \
     wget
 
-RUN wget -qO - http://download.opensuse.org/repositories/home:/laszlo_budai:/syslog-ng/Debian_8.0/Release.key | apt-key add -
-RUN echo 'deb http://download.opensuse.org/repositories/home:/laszlo_budai:/syslog-ng/Debian_8.0 ./' | tee --append /etc/apt/sources.list.d/syslog-ng-obs.list
+RUN wget -qO - \
+http://download.opensuse.org/repositories/home:\
+/laszlo_budai:/syslog-ng/Debian_8.0/Release.key | apt-key add -
```
Comment (lvh) [inline script/syslog-ng/Dockerfile]: I don' t know who laszlo is but I support him having root on my images

Comment (lvh) [inline script/syslog-ng/Dockerfile]: (I appreciate that you didn't change this code.)

File: resources/test/index.html (hunk 10 above)
Comment (sirsean) [inline resources/test/index.html]: Because it had been in a single line in the unit test that verified loading the index, and there was no way to break it up into multiple lines to pass eclint without breaking the test.

File: script/syslog-ng/Dockerfile (hunk 11 above)
Comment (sirsean) [inline script/syslog-ng/Dockerfile]: I debated omitting Dockerfiles from eclint. Decided not to. :-)

File: src/desdemona/plugins/swift.clj [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+  (find-first #(and (= "cloudFiles" (:name %)) (= "object-store" (:type %)))
+              (-> auth-response :access :service-catalog)))
+
+(defn get-token
+  "Reach into an authentication response and extract the auth-related part
+  of it (which will include both the auth token and the expiration date)."
+  [auth-response]
+  (-> auth-response :access :token))
+
+(defn authenticate
```
Comment (lvh) [inline src/desdemona/plugins/swift.clj]: This is pretty side-effecty; it should probably end in `!`

File: src/desdemona/plugins/swift.clj [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+  (let [body (json/generate-string {"auth" {"RAX-KSKEY:apiKeyCredentials"
+                                            {"username" username
+                                             "apiKey" api-key}}})
+        response (http/post auth-url
+                            {:body body
+                             :content-type :json
+                             :accept :json
+                             :throw-entire-message true
+                             :as :stream})
+        decoded (json/decode-stream (bs/to-reader (response :body))
```
Comment (lvh) [inline src/desdemona/plugins/swift.clj]: Since you're just going to return this verbatim; consider making it the "body" of the let. Also, that expr would probably look a little nicer with `->`, but that's up to you

**Comment by lvh** [comment]
The docker build appears to report the tests 4 times: https://jenkins.racksec.io/job/desdemona-docker-pr-builder/22/console

... but the new swift plugin tests don't appear to be a part of that.

File: project.clj (hunk 9 above)
Comment (lvh) [inline project.clj]: Aleph (and manifold, the library it builds on), are the libs we (esp @derwolfe and I) currently use most. It has a number of awesome features; notably asynchronous IO support in a way that is compatible with many different other libs, explicit backpressure, &c. I'm not sure it's worth switching; I'll read through the rest of the PR and think about it.

File: src/desdemona/plugins/swift.clj [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+  p-ext/Pipeline
+
+  (read-batch
+    [_ event]
+    (function/read-batch event))
+
+  (write-batch
+    [_ {:keys [onyx.core/results]}]
+    (let [segments (map :message (mapcat :leaves (:tree results)))]
+      (when (not-empty segments)
```
Comment (lvh) [inline src/desdemona/plugins/swift.clj]: map returns nil xor a seq; so you can use when-let here

**Comment by sirsean** [comment]
That is ... weird.

```Ran 38 tests containing 141 assertions.```

That's what I see locally, and the Docker build is clearly not doing that. I'll look into it.

File: src/desdemona/plugins/swift.clj [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+    {}))
+
+(defn write-batch
+  "This is the function that gets added to the catalog, and sets up the plugin
+  that actually writes batches to Cloud Files."
+  [pipeline-data]
+  (let [task-map (:onyx.core/task-map pipeline-data)
+        auth-url (:swift/auth-url task-map)
+        username (:swift/username task-map)
+        api-key (:swift/api-key task-map)]
```
Comment (lvh) [inline src/desdemona/plugins/swift.clj]: did destructuring not work out well for this? (I'd even consider juxt, but I'm an animal)

File: src/desdemona/plugins/swift.clj (hunk 14 above)
Comment (lvh) [inline src/desdemona/plugins/swift.clj]: (also as a general rule don't use empty/not-empty when simple nil/not-nil will do, e.g. you should almost always use `(seq x)` in favor of `(not (empty? x))` or `(not-empty x)`)

**Comment by sirsean** [comment]
Still can't figure out what's going on with the Docker tests in CI. It runs all the tests for me (even inside Docker locally) but clearly isn't running the new tests in CI. The `docker-compose` file shows that the command it's running is just `lein test`, which should really work.

File: script/mimic (hunk 4 above)
Comment (reaperhulk) [inline script/mimic]: This isn't a proper submodule since there's no `.gitmodules` so it can't be fetched from the branch. This causes the mimic docker image to fail building (which you can see in the output on jenkins).

Comment (reaperhulk) [inline script/mimic]: If you want to test that this works as expected you'll want to clone a new copy of the repo and then runs these commands:

```
git fetch origin +refs/pull/129/merge:
git checkout -qf FETCH_HEAD
git submodule update --init --recursive
```

If mimic exists in `script/mimic` after that you're good.
